### ISSUE TITLE
[Question] Hitting new Docker Hub rate limit

### ISSUE THREAD

**Issue by Karnak19** [caprover/caprover#906]
Hello,

Since today, I have an issue with the newly added Docker Hub rate limit.

I have like 4 apps on my Caprover instancce, all of them are already built and online.

But still, when I curl the rate limit preview endpoint on docker.io, it looks like something is constantly trying to pull.

https://www.docker.com/blog/checking-your-current-docker-pull-rate-limits-and-status/

```sh
TOKEN=$(curl "https://auth.docker.io/token?service=registry.docker.io&scope=repository:ratelimitpreview/test:pull" | jq -r .token)
```

```sh
curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
```

output:

```sh
root@ubuntu-s-1vcpu-1gb-fra1-01:~# curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 100;w=21600
RateLimit-Remaining: -472;w=21600
root@ubuntu-s-1vcpu-1gb-fra1-01:~# curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 100;w=21600
RateLimit-Remaining: -482;w=21600
root@ubuntu-s-1vcpu-1gb-fra1-01:~# curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 100;w=21600
RateLimit-Remaining: -512;w=21600
root@ubuntu-s-1vcpu-1gb-fra1-01:~# curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 100;w=21600
RateLimit-Remaining: -518;w=21600
root@ubuntu-s-1vcpu-1gb-fra1-01:~# curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 100;w=21600
RateLimit-Remaining: -526;w=21600
root@ubuntu-s-1vcpu-1gb-fra1-01:~# curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
root@ubuntu-s-1vcpu-1gb-fra1-01:~# curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 100;w=21600
RateLimit-Remaining: -535;w=21600
```

**Comment by gregflint** [comment]
I found that one of my applications (I have way more than four) was failing and restarting and each time it restarted it seemed to try to pull a new image from docker (no surprise there). That's what led to me pulling more than 100 times in a six hour period.  Once I set that app instance count to zero everything started to stabilize. Not a "solution" but I hope that helps.

**Comment by githubsaturn** [comment]
CapRover doesn't pull images randomly on the background. You can temporarily shutdown CapRover to see if it has any effects

```
docker service scale captain-captain=0
```

**Comment by shashankklenty** [comment]
I am currently facing the same issue. It seems after performing the `docker login` command on the caprover host, caprover's swarm service creation does not use the credentials stored locally.

**Comment by githubsaturn** [comment]
CapRover uses Docker API to perform actions. When you perform `docker login` you are logging in with your CLI, but any apps (like CapRover) that uses Docker API on your host is unable to access the credentials.

The workaround for this is quite intensive:
1- User muse login to DockerHub using CapRover. Cluster > Add Remote Registry. `URL: index.docker.io` username... password
2- We need to make some changes to the code to send credentials to DockerHub, even for public images.

**Comment by shashankklenty** [comment]
@githubsaturn thanks so much for the quick response. So i followed step 1, when setting url to `index.docker.io` it was giving me an authentication error however when i switched to `registry-1.docker.io` it successfully authenticated and now my OCA builds are working! :)

**Comment by githubsaturn** [comment]
So no changes were needed in the one click apps? It just worked? That surprises me, but a good surprise!

If that's the case, we can close this issue then.

**Comment by shashankklenty** [comment]
I guess not, it just worked :joy: If i do encounter anything I'll be back :stuck_out_tongue: Thanks again! :smiley:

**Comment by mdehollander** [comment]
I am also running into the newly set rate limits, which seems to be active since yesterday. I creating a template for a new app, and only pulling postgres. On my caprover instance is only 1 other app.

In `syslog` you can see:
```
Nov 21 10:46:46 caprover dockerd[14539]: time="2020-11-21T10:46:46.748300640+01:00" level=error msg="Not continuing with pull after error: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit"
```
And checking the rate limit with the command from the first message:
```
$ curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 100;w=21600
RateLimit-Remaining: 0;w=21600
```

I don't have swarm/cluster setup or use a different registry. What is a solution for the rate limit? I don't think I have been pulling 100 times, because I have only uploaded a template for development approx 8 times or so. Thanks for any advice.

**Comment by githubsaturn** [comment]
As mentioned above, CapRover it self does not pull any images. Here is the summary of what people have tried:
* Make sure your Docker CLI has access to your token. Do this by running `docker login`
* Make sure CapRover engine has access to your token:
```
- Go to cluster tab
- Add remote registry
- URL: registry-1.docker.io    username: yourusername    imageprefix: yourusername     password: yourpassword
- Save
- IMPORTANT: Make sure you DISABLE PUSH if you don't have a cluster of nodes
```
With these steps both CLI and CapRover will follow the authenticated limits which is a bit higher than anonymous limit.
Also, keep in mind that an application failure can force Docker to keep pulling the image: https://github.com/caprover/caprover/issues/906#issuecomment-728300639

![image](https://user-images.githubusercontent.com/10582082/99877843-ffcfd500-2bce-11eb-975e-a3a28e0107da.png)

**Comment by kgnfth** [comment]
> As mentioned above, CapRover it self does not pull any images. Here is the summary of what people have tried:
> 
> * Make sure your Docker CLI has access to your token. Do this by running `docker login`
> * Make sure CapRover engine has access to your token:
> 
> ```
> - Go to cluster tab
> - Add remote registry
> - URL: registry-1.docker.io    username: yourusername    imageprefix: yourusername     password: yourpassword
> - Save
> - IMPORTANT: Make sure you DISABLE PUSH if you don't have a cluster of nodes
> ```
> 
> With these steps both CLI and CapRover will follow the authenticated limits which is a bit higher than anonymous limit.
> Also, keep in mind that an application failure can force Docker to keep pulling the image. It's not done by CapRover, it's done by Docker: [#906 (comment)](https://github.com/caprover/caprover/issues/906#issuecomment-728300639)
> 
> ![image](https://user-images.githubusercontent.com/10582082/99877843-ffcfd500-2bce-11eb-975e-a3a28e0107da.png)

this does not work for me 
i still get the same rate limit error

**Comment by mdehollander** [comment]
At first it also did not work for me after I create a login at docker.com. But after a day (maybe when the limit was reset) the same tasks and more where possible to perform. So maybe a reset of the limit is needed first, before the new authenticated limit starts?

**Comment by kgnfth** [comment]
> At first it also did not work for me after I create a login at docker.com. But after a day (maybe when the limit was reset) the same tasks and more where possible to perform. So maybe a reset of the limit is needed first, before the new authenticated limit starts?

not sure but i have this when i authenticate using cli

❯ curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 200;w=21600
RateLimit-Remaining: 200;w=21600

**Comment by Scarzer** [comment]
Hiya, I'm now three days into this issue. I've re-authed both via docker CLI and the caprover interface following the instructions on this issue. I'm still stuck at 100 requests, and I somehow always have 0 requests remaining. This is even after giving it 24 hours to reset...

I'm at a loss, every time I try to deploy an application I get a 500 error.

**Comment by kgnfth** [comment]
@Scarzer check your rate limit https://docs.docker.com/docker-hub/download-rate-limit/#how-can-i-check-my-current-rate

**Comment by Scarzer** [comment]
@kgnfth I have... I have 0 remaining requests. It's been that way for 3 days now. I'm going through my applications one by one now turning them off to make sure they aren't making requests in the background.

**Comment by kgnfth** [comment]
so you have 0 remaining with a logged in user ?

**Comment by Scarzer** [comment]
@kgnfth Yes. 

I started getting 500 errors 3 days ago when I started to check out your latest one click apps (thx btw). I came here and found this thread. Checked my rate limit and saw that I had 0 remaining out of 100. I logged in to docker hub via the CLI on my server. I added docker registry to my caprover server as well. I turned off push services, I'm not using a cluster. 

My errors persisted. For 3 days.

**Comment by Scarzer** [comment]
```
root@caprover:~# TOKEN=$(curl "https://auth.docker.io/token?service=registry.docker.io&scope=repository:ratelimitpreview/test:pull" | jq -r .token)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4369    0  4369    0     0  80907      0 --:--:-- --:--:-- --:--:-- 80907

root@caprover:~# curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 100;w=21600
RateLimit-Remaining: 0;w=21600
```
I am 100% sure I am logged in via the CLI. Every time I login, I can see the timestamp and IP address on the login token being used. Even if I wasn't a logged in user, I absolutely not using 100 requests when I'm sitting around and waiting for the rate limit to reset. 

Additionally, if I pull down some random image using the docker cli, it works without any issues.

**Comment by kgnfth** [comment]
@Scarzer You are not logged in because anonymous usage has 100 limit while free user has 200,
you have to run it like this

replace username:password 

1. `TOKEN=$(curl --user 'username:password' "https://auth.docker.io/token?service=registry.docker.io&scope=repository:ratelimitpreview/test:pull" | jq -r .token)`

2. `curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit`

**Comment by Scarzer** [comment]
@kgnfth Thank you for the hint to authenticate the token request. It deepens the mystery because the credentials saved by the docker CLI are not being used when requesting a token otherwise... Caprover is still throwing 500 errors and my authentication details have not changed in 3 days.

**Comment by githubsaturn** [comment]
Try running the line below. It adds your docker registry to all local services.

```
curl -sSL https://raw.githubusercontent.com/caprover/caprover/master/dev-scripts/add-registry-to-all-services.sh | bash -s -- 
```

Then wait for an hour.

**Comment by Scarzer** [comment]
As was mentioned in [#906(comment)](https://github.com/caprover/caprover/issues/906#issuecomment-728300639) I had a failing postgresdb constantly attempting to pull down an image but failed to do so.... I can deploy again!! =D 

My API limit is still 100 requests so I'm going to keep trying to figure out why I'm unable to authenticate. I'm trying what you linked too @githubsaturn and will report back if my authentication issues stop.

**Comment by kgnfth** [comment]
@githubsaturn i was just testing some custom apps but i just discovered that caprover did not use my registry which i added 5 days ago
i have 0 anonymous limit
and 183 with my username

![image](https://user-images.githubusercontent.com/73314940/100524117-af034200-31b5-11eb-9e15-034e5b58fae3.png)

**Comment by githubsaturn** [comment]
@kgnfth  - I just [pushed some changes](https://github.com/caprover/caprover/commit/9094d116635a44ef4f596e426cf2486310937428) to explicitly force usage of docker hub registry. 

Wait for 1hr for the build to finish and then migrate to [edge version](https://caprover.com/docs/troubleshooting.html#how-to-use-the-edge-version) to see if it resolves the issue. Let me know if it still doesn't work

**Comment by kgnfth** [comment]
@githubsaturn Yes it works
Before
```
❯ curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 200;w=21600
RateLimit-Remaining: 182;w=21600
```
After
```
❯ curl --head -H "Authorization: Bearer $TOKEN" https://registry-1.docker.io/v2/ratelimitpreview/test/manifests/latest 2>&1 | grep RateLimit
RateLimit-Limit: 200;w=21600
RateLimit-Remaining: 176;w=21600
```

**Comment by githubsaturn** [comment]
Great - it'll be part of the next release. For now you can stay on the edge image.

**Comment by githubsaturn** [comment]
Update - originally suggested here: https://github.com/caprover/caprover/issues/935#issuecomment-743061499

TLDR: 
You can use a cache mirror that helps with rate limits even with unauthenticated requests:
```
echo '{"registry-mirrors": ["https://mirror.gcr.io"]}' > /etc/docker/daemon.json
sudo service docker restart
```

**Comment by getaway44** [comment]
Hello,

I just wanted to let you know that switching to the edge version was not fixing the issue for me (I have the exact same issue than kgnfth). It keeps pulling image without registration to docker (in anonymous mode)

Thanks for your workaround with the cache mirror. If I understood well, we wont need this workaround in the next caprover version. Do you know when the next version will be released ? 

Thank you for your work and for the time you spend on answering ont Github issues. 👍

**Comment by githubsaturn** [comment]
> If I understood well, we wont need this workaround in the next caprover version.

Next version of CapRover is just "edge". If edge doesn't work for you, the new version won't be of any help. But the cached mirror solution doesn't require a CapRover update, it's low level with Docker. Try it out, and if it works, it's great.

**Comment by crypdick** [comment]
I'm in a catch-22 where a bunch of services are throwing 502 errors but I can't fix them because of this issue (multiple apps are dependent on 1 app, which is broken until I pull a new image, but I can't pull the image because of rate-limiting). Is there a way to "pause" the broken apps so that I can reset my limit? It would really suck if I had to delete the services.

**Comment by githubsaturn** [comment]
Run this to pause all services:
```
all_services=$(docker service ls --format {{.Name}})
for srv in $all_services
do
  echo "Service: $srv"
  docker service scale $srv=0 -d
done
```
Of course this will also stop CapRover. If you want to keep caprover running:
```
  docker service scale captain-captain=0 -d
  docker service scale captain-nginx=0 -d
  docker service scale captain-certbox=0 -d
```



After you fixed the root cause and you want to restore all your services:
```
all_services=$(docker service ls --format {{.Name}})
for srv in $all_services
do
  echo "Service: $srv"
  docker service scale $srv=1 -d
done
```

**Comment by efritze** [comment]
I recently faced this issue.

None of the above worked for me, even though I had a Pro Dockerhub account.

It appears Caprover (v1.9.0) was not properly using my Dockerhub Pro credentials for authentication, which wasn't an issue before. 

After some trial and error, this is what worked for me:

1. Enter your Dockerhub credentials inside Caprover (Caprover -> Cluster -> Add Remote Registry).  Use `registry.hub.docker.com`  as domain name. If you previously used registry-1.docker.io as domain name, change it to registry.hub.docker.com
2. In your deployment scripts, prefix your image names with  `registry.hub.docker.com`.  So instead of deploying the docker image: myinc/myapp, deploy  registry.hub.docker.com/myinc/myapp

**Comment by githubsaturn** [comment]
Looks like there are multiple aliases for the registry domain. And people use [different ones](https://github.com/moby/moby/issues/20804#issuecomment-191912057):
```
registry-1.docker.io
registry.hub.docker.com
docker.io
index.docker.io
```

**Comment by real34** [comment]
If it can help someone (or a future me), having Gitlab runners on your Caprover may lead to reaching the limit easily (each job would pull an image).

There's a Gitlab feature flag to use their docker registry and prevent consuming your instance's limits: see https://docs.gitlab.com/runner/configuration/advanced-configuration.html#migrate-helper-image-to-registrygitlabcom
